File: guessing-game-bdd-atomic/src/test/java/tw/waterballsa/bddworkshop/stepdefs/api/API_He_Is_Greeting_With.java
package tw.waterballsa.bddworkshop.stepdefs.api;

import io.cucumber.datatable.DataTable;
import io.cucumber.java.en.When;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.MediaType;
import org.springframework.test.web.servlet.MockMvc;
import org.springframework.test.web.servlet.MvcResult;
import tw.waterballsa.bddworkshop.stepdefs.ScenarioContext;

import java.util.Map;

import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.get;

public class API_He_Is_Greeting_With {

    @Autowired
    private MockMvc mockMvc;

    @Autowired
    private ScenarioContext scenarioContext;

    @When("he is greeting, with:")
    public void invoke(DataTable dataTable) throws Exception {
        // 1. Extract data from DataTable
        Map<String, String> data = dataTable.asMaps().get(0);
        String name = data.get("name");

        // 2. Call the API according to the API spec
        MvcResult result = mockMvc.perform(get("/greeting")
                .param("name", name)
                .contentType(MediaType.APPLICATION_JSON))
                .andReturn();

        // 3. Store the response in the ScenarioContext
        scenarioContext.setLastResponse(result);
    }
}

File: guessing-game-bdd-atomic/src/test/stepdefs/api/API_He_Is_Greeting_With.ts
import { When } from '@cucumber/cucumber';
import { DataTable } from '@cucumber/cucumber';
import * as request from 'supertest';
import { ScenarioContext } from '../ScenarioContext';
import { INestApplication } from '@nestjs/common';

export class API_He_Is_Greeting_With {
    constructor(
        private readonly app: INestApplication,
        private readonly scenarioContext: ScenarioContext
    ) {}

    @When('he is greeting, with:')
    public async invoke(dataTable: DataTable): Promise<void> {
        // 1. Extract data from DataTable
        const data = dataTable.hashes()[0];
        const name = data.name;

        // 2. Call the API according to the API spec
        const response = await request(this.app.getHttpServer())
            .get('/greeting')
            .query({ name })
            .set('Content-Type', 'application/json')
            .expect(200);

        // 3. Store the response in the ScenarioContext
        this.scenarioContext.setLastResponse(response);
    }
}
